File: oauth2/src/jvmTest/java/com/okta/oauth2/kmp/jvm/RedirectEndSessionFlowTest.java
/*
 * Copyright 2022-Present Okta, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.okta.oauth2.kmp.jvm;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import com.okta.oauth2.kmp.BrowserRedirectHandler;
import java.util.Collections;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import kotlin.Unit;
import org.junit.Test;

/**
 * Java-language tests for {@link RedirectEndSessionFlow} CompletableFuture wrapper. Validates that
 * the API is ergonomic from actual Java code.
 */
public class RedirectEndSessionFlowTest {

  @Test
  public void start_ReturnsCompletableFutureOnSuccess()
      throws ExecutionException, InterruptedException, TimeoutException {
    BrowserRedirectHandler handler =
        TestFlowFactory.createFakeBrowserRedirectHandler(
            "com.example.app:/logout?state=test-state");
    RedirectEndSessionFlow flow =
        TestFlowFactory.createSuccessRedirectEndSessionFlow("com.example.app:/logout");
    try {
      CompletableFuture<Unit> future =
          flow.start("example-id-token", "com.example.app:/logout", handler);
      assertNotNull("Future should not be null", future);
      future.get(5, TimeUnit.SECONDS);
    } finally {
      flow.close();
    }
  }

  @Test
  public void start_WithAllParams_ReturnsUnit()
      throws ExecutionException, InterruptedException, TimeoutException {
    BrowserRedirectHandler handler =
        TestFlowFactory.createFakeBrowserRedirectHandler(
            "com.example.app:/logout?state=test-state");
    RedirectEndSessionFlow flow =
        TestFlowFactory.createSuccessRedirectEndSessionFlow("com.example.app:/logout");
    try {
      CompletableFuture<Unit> future =
          flow.start(
              "example-id-token", "com.example.app:/logout", handler, Collections.emptyMap());
      future.get(5, TimeUnit.SECONDS);
    } finally {
      flow.close();
    }
  }

  @Test
  public void start_WithError_CompletesExceptionally()
      throws TimeoutException, InterruptedException {
    BrowserRedirectHandler handler =
        TestFlowFactory.createFakeBrowserRedirectHandler("com.example.app:/logout");
    RedirectEndSessionFlow flow = TestFlowFactory.createFailingRedirectEndSessionFlow();
    try {
      CompletableFuture<Unit> future =
          flow.start("example-id-token", "com.example.app:/logout", handler);
      try {
        future.get(5, TimeUnit.SECONDS);
        fail("Should have thrown ExecutionException");
      } catch (ExecutionException e) {
        assertNotNull("Cause should not be null", e.getCause());
        assertTrue(
            "Should contain error message",
            e.getCause().getMessage().contains("OIDC Endpoints not available."));
      }
    } finally {
      flow.close();
    }
  }

  @Test
  public void close_IsIdempotent() {
    RedirectEndSessionFlow flow =
        TestFlowFactory.createSuccessRedirectEndSessionFlow("com.example.app:/logout");
    flow.close();
    flow.close(); // Should not throw
  }
}

File: oauth2/src/commonTest/kotlin/com/okta/oauth2/kmp/RedirectEndSessionFlowTest.kt
/*
 * Copyright 2022-Present Okta, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.okta.oauth2.kmp

import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class RedirectEndSessionFlowTest {
    @Test
    fun start_WhenSuccess_ReturnsFlowContext() =
        runTest {
            val mockFlow =
                object : RedirectEndSessionFlow {
                    override suspend fun start(
                        idToken: String,
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                    ): Result<RedirectEndSessionFlowContext> =
                        Result.success(
                            RedirectEndSessionFlowContext(
                                url = "https://example.okta.com/logout?id_token_hint=$idToken",
                                redirectUrl = redirectUrl,
                                state = "test-state"
                            )
                        )

                    override fun resume(
                        uri: String,
                        flowContext: RedirectEndSessionFlowContext,
                    ): Result<Unit> = Result.success(Unit)
                }

            val result = mockFlow.start("test-id-token", "com.example.app:/logout")

            assertTrue(result.isSuccess)
            val context = result.getOrNull()
            assertNotNull(context)
            assertEquals("com.example.app:/logout", context.redirectUrl)
            assertTrue(context.url.contains("test-id-token"))
        }

    @Test
    fun resume_WhenSuccess_ReturnsUnit() =
        runTest {
            val mockFlow =
                object : RedirectEndSessionFlow {
                    override suspend fun start(
                        idToken: String,
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                    ): Result<RedirectEndSessionFlowContext> =
                        Result.success(
                            RedirectEndSessionFlowContext(
                                url = "https://example.okta.com/logout",
                                redirectUrl = "com.example.app:/logout",
                                state = "test-state"
                            )
                        )

                    override fun resume(
                        uri: String,
                        flowContext: RedirectEndSessionFlowContext,
                    ): Result<Unit> = Result.success(Unit)
                }

            val context = mockFlow.start("id-token", "com.example.app:/logout").getOrThrow()
            val result = mockFlow.resume("com.example.app:/logout?state=test-state", context)

            assertTrue(result.isSuccess)
        }

    @Test
    fun resume_WhenEndpointUnavailable_ReturnsFailure() =
        runTest {
            val mockFlow =
                object : RedirectEndSessionFlow {
                    override suspend fun start(
                        idToken: String,
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                    ): Result<RedirectEndSessionFlowContext> = Result.failure(IllegalStateException("OIDC Endpoints not available."))

                    override fun resume(
                        uri: String,
                        flowContext: RedirectEndSessionFlowContext,
                    ): Result<Unit> = Result.success(Unit)
                }

            val result = mockFlow.start("id-token", "com.example.app:/logout")

            assertTrue(result.isFailure)
            assertTrue(result.exceptionOrNull()?.message?.contains("OIDC Endpoints not available.") == true)
        }

    @Test
    fun resume_WhenStateMismatch_ReturnsFailure() =
        runTest {
            val mockFlow =
                object : RedirectEndSessionFlow {
                    override suspend fun start(
                        idToken: String,
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                    ): Result<RedirectEndSessionFlowContext> =
                        Result.success(
                            RedirectEndSessionFlowContext(
                                url = "https://example.okta.com/logout",
                                redirectUrl = "com.example.app:/logout",
                                state = "expected-state"
                            )
                        )

                    override fun resume(
                        uri: String,
                        flowContext: RedirectEndSessionFlowContext,
                    ): Result<Unit> =
                        if (uri.contains("wrong-state")) {
                            Result.failure(RedirectEndSessionFlow.ResumeException("Failed due to state mismatch."))
                        } else {
                            Result.success(Unit)
                        }
                }

            val context = mockFlow.start("id-token", "com.example.app:/logout").getOrThrow()
            val result = mockFlow.resume("com.example.app:/logout?state=wrong-state", context)

            assertTrue(result.isFailure)
            assertTrue(result.exceptionOrNull()?.message?.contains("state mismatch") == true)
        }

    @Test
    fun resume_WhenErrorReturned_ReturnsFailure() =
        runTest {
            val mockFlow =
                object : RedirectEndSessionFlow {
                    override suspend fun start(
                        idToken: String,
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                    ): Result<RedirectEndSessionFlowContext> =
                        Result.success(
                            RedirectEndSessionFlowContext(
                                url = "https://example.okta.com/logout",
                                redirectUrl = "com.example.app:/logout",
                                state = "test-state"
                            )
                        )

                    override fun resume(
                        uri: String,
                        flowContext: RedirectEndSessionFlowContext,
                    ): Result<Unit> = Result.failure(RedirectEndSessionFlow.ResumeException("access_denied"))
                }

            val context = mockFlow.start("id-token", "com.example.app:/logout").getOrThrow()
            val result = mockFlow.resume("com.example.app:/logout?error=access_denied", context)

            assertTrue(result.isFailure)
            assertNotNull(result.exceptionOrNull())
        }

    @Test
    fun start_WithExtraParameters_IncludesThemInContext() =
        runTest {
            var capturedExtraParams: Map<String, String>? = null
            val mockFlow =
                object : RedirectEndSessionFlow {
                    override suspend fun start(
                        idToken: String,
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                    ): Result<RedirectEndSessionFlowContext> {
                        capturedExtraParams = extraRequestParameters
                        return Result.success(
                            RedirectEndSessionFlowContext(
                                url = "https://example.okta.com/logout",
                                redirectUrl = redirectUrl,
                                state = "test-state"
                            )
                        )
                    }

                    override fun resume(
                        uri: String,
                        flowContext: RedirectEndSessionFlowContext,
                    ): Result<Unit> = Result.success(Unit)
                }

            mockFlow.start(
                idToken = "test-id-token",
                redirectUrl = "com.example.app:/logout",
                extraRequestParameters = mapOf("ui_locales" to "en")
            )

            assertNotNull(capturedExtraParams)
            assertEquals("en", capturedExtraParams["ui_locales"])
        }
}
